R-package/src/install.libs.R

```diff
@@ -14,33 +14,19 @@ if (!(R_int_UUID == "0310d4b8-ccb1-4bb8-ba94-d36a55f60262"
     || R_int_UUID == "2fdf6c18-697a-4ba7-b8ef-11c0d92f1327")){
   print("Warning: unmatched R_INTERNALS_UUID, may cannot run normally.")   
 }
+
+# Move in CMakeLists.txt
+if (!file.copy("../inst/bin/CMakeLists.txt", "CMakeLists.txt", overwrite = TRUE)){
+  stop("Copying CMakeLists failed")
+}
+
 # Check for precompilation
 if (!use_precompile) {
 
   # Check repository content
   source_dir <- file.path(R_PACKAGE_SOURCE, "src", fsep = "/")
   setwd(source_dir)
   
-  if (!file.exists("_IS_FULL_PACKAGE")) {
-    unlink("./include", recursive = TRUE)
-    unlink("./src", recursive = TRUE)
-    unlink("./compute", recursive = TRUE)
-    unlink("./build", recursive = TRUE)
-    if (!file.copy("./../../include", "./", overwrite = TRUE, recursive = TRUE)) {
-      stop("Cannot find folder LightGBM/include")
-    }
-    if (!file.copy("./../../src", "./", overwrite = TRUE, recursive = TRUE)) {
-      stop("Cannot find folder LightGBM/src")
-    }
-    if (!file.copy("./../../compute", "./", overwrite = TRUE, recursive = TRUE)) {
-      print("Cannot find folder LightGBM/compute, disabling GPU build.")
-      use_gpu <- FALSE
-    }
-    if (!file.copy("./../../CMakeLists.txt", "./", overwrite = TRUE, recursive = TRUE)) {
-      stop("Cannot find file LightGBM/CMakeLists.txt")
-    }
-  }
-  
```

Comment:
why remove these ?

Reply:
@guolinke having these inside the package only works if you have the entire repo laying around (since they reach back up outside of `R-package` to the repo root).

That will always fail for people installing the tarball from CRAN. Instead, we should follow the approach that I laid out in the description where we copy the source files into `R-package/` in the appropriate place, and then bundle them with the package tarball.

Does that make sense? I'm thinking I should PR in that process (with all the `cp` calls as a build script in this repo)

Reply:
So there is "_full_package" and "build_package.R" . After run "build_package.R", there is a full package with tarball format.

Reply:
Having the "copy around all the source files" logic in two places was confusing me and making it difficult to debug when I accidentally took some action from the wrong directory. I think that this is much more straightforward.

Also, like I said the calls inside this `install.libs.R` will NOT work when a user downloads the package tarball from CRAN (since they assume that you can reach back up to e.g. the `include/` directory at the repo root).

Whether we keep `build_package.R` (with some modifications like the `--keep-empty-dirs flag`) or use the shell script I proposed, I think this logic inside `install.libs.R` and the `_full_package` thing should be killed.

Reply:
@jameslamb the python-package share the same logic. BTW, it will always copy the files from root folder, so you can always edit the file in root folder, not the R-package. 
Cannot we push the tar file after call "build_package.R" to CRAN ? 
ping @Laurae2 .

Reply:
I did not edit any of the source files in the root, just the way we get those files into the R-package.

Reply:
We cannot precompile and distribute a binary package via R (the implication of "running build_package.R and pushing to CRAN"). See the CRAN policies here: https://cran.r-project.org/web/packages/policies.pdf, esp.

```
Binary packages are not accepted from maintainers
```

And

```
Source packages may not contain any form of binary executable code
```

We have to push a *source* package to CRAN, which will get compiled on install when users install with `install.packages`. When that happens, the code inside `install.libs.R` will run on the user's machine. At that point, they will not have access to the repo root so having code inside `install.libs.R` that expects to be able to reach back above `R-package` does not work.

We *can* just run `build_package.R` instead of the shell script approach I proposed if you feel more comfortable with that, but either way this `_full_package` thing and the if block in question in this diff need to go away before we can push to CRAN.

Reply:
build_package.R is building the source package, not the binary one. 
It copy the include and src to the R-package folder, and pack it to tar file.

Reply:
For sure, I can help with the R & CRAN and I hope @Laurae2 will agree with the approach I've taken. I agree that `build_package.R` is building the source package...what I'm saying is that the block of logic inside `(!file.exists("_IS_FULL_PACKAGE")` will never work outside of the case where you are building with this repo. So it should be removed before this goes to CRAN.

@Laurae2 and I will discuss

Reply:
@guolinke @jameslamb the `is_full_package` logic in R should be removed if the package gets pushed to CRAN.

But - we could also do a script which setups the package specifically for CRAN like xgboost does. Those lines could be removed using that script for instance (I think this is what @jameslamb meant in the 2nd comment) => it won't break existing repositories depending on the current status of LightGBM, and it won't break for CRAN.

Reply:
I really want to just kill the `is_full_package` logic completely. There's no reason for it to exist. It doesn't make sense to have two different build processes, and it was making it very difficult to develop and test.

If you agree, I will PR in my build script (the one I mentioned in the description) and update the "how to build the R package" instructions.

Do either of you have comments on the rest of the changes I made in this PR? I think we are very close to being CRAN-able and would love to make forward progress soon.

Reply:
@jameslamb I'm OK

@guolinke any opinion?

Reply:
As long as the installation is easy both from github/cran on all platforms (osx, linux, windows), I am okay with that.

Reply:
Ok great, thank you. I owe it to both of you to update the R install instructions and check that this works on Windows. I'll report back when I've done that.

Thanks for taking the time to review. Hope you're as excited as I am that CRAN is close!!

Reply:
@jameslamb I'll check tomorrow for installing in Windows (MinGW, Visual Studio 2015, Visual Studio 2017, Intel Compilers) and WSL (gcc).

Reply:
Thank you @Laurae2 . I have been trying to test on my windows laptop, but need to update some things to get it running. Looking forward to hearing what you say.

It's possible that we will want to stick with the concept of an R script for building (since using `devtools` functions makes it cross-OS for free) but I do still feel strongly about removing the `is_full_package` logic.

Also, in case I didn't say it before, I will hold off on fixing that remaining note about asserts and logs until we merge this PR. I'll just copy the way XGBoost did it, but don't want to add anything else to this already large PR.

Reply:
@Laurae2 have you had a chance to test this?

Reply:
@jameslamb I'm running it right now.

That CMakeLists.txt is really a killer at the moment, should we keep it in install.libs.R? Always getting an error here without your `cp CMakeLists.txt R-package/inst/bin/` fix:

https://github.com/jameslamb/LightGBM/blob/bugfix/cran_stuff/R-package/src/install.libs.R#L18-L21

Reply:
sorry, getting back to this. I feel strongly that `install.libs.R` shouldn't know anything about shared files from the repo root (i.e. things outside the R package) and that my proposal for a `build_r.sh` should be the preferred way. I'll add that file to the PR and to the R install instructions.

Or maybe it should be back to something like `build_package.R` so I don't have to deal with cross-OS issues? Can test that on my laptop in the next couple days.

I'll also rebase to current master, know this branch has fallen behind